captum/optim/models/__init__.py

```diff
@@ -0,0 +1,6 @@
+from ._common import RedirectedReluLayer  # noqa: F401
```

Comment:
You should also add `skip_layers` here as well

Reply:
It looks like we are exposing `_core` twice ? I think that we should expose oly specific classes, functions from `_core`.

```
from captum.optim._core import loss  # noqa: F401
from captum.optim._core import optimization  # noqa: F401
```

Reply:
@NarineK do you mean something like this?
```
from captum.optim._core import loss, optimization  # noqa: F401
```

Reply:
I think this way we are exposing everything in loss and optimization files. One day we might have some classes there that we don't want to expose to users. I think it would be good to define `__all__`  similar to this: 
https://github.com/pytorch/pytorch/blob/master/torch/__init__.py#L34 and also see captum attribution.
 https://github.com/pytorch/captum/blob/optim-wip/captum/attr/__init__.py#L53
In __all__ we can list all public classes and functions that we want to expose to users.

Reply:
Oh yeah sure I can add that. In that case will we be doing something like this instead?

i.e. users will access the losses etc. directly from `captum.optim` instead of `captum.optim.loss`.

```
from captum.optim._core.loss import (
    BaseLoss,
    CompositeLoss,
    ...
)

__all__ = [
    "BaseLoss",
    "CompositeLoss",
    ...
]
```

Reply:
Yeah, I think that being more explicit will help us to know which ones are exactly in the public API.

Reply:
@NarineK can you see if the latest commit does what you intended?

Although @ProGamerGov and I do kinda prefer the previous __init__ because it feels a little more structured e.g. users access losses via `captum.optim.loss` and image-parameterizations via `captum.optim.image`.

Reply:
@greentfrapp, I think @NarineK might have meant that we should filter the submodules like `transform` and `image` so that we didn't just import everything from them, and not remove them from the user path. As the optim module has many different submodules, it could be confusing to users if we remove them from the the user import paths. Also, things would only get more confusing once we place the appropriate submodules under the `optim.vis` submodule to separate them from non visual optimizations.

Reply:
I liked that the users import it like: `from captum.optim.loss import L2`. I agree that it is better structured. In order to make the API clear I wanted to list them in `__all__`. In `__all__` we kind of define a public API. It doesn't prohibit users to use other things from loss or transform etc. But it is nice for the user.
Some links about `__all__`:
https://stackoverflow.com/questions/2187583/whats-the-python-all-module-level-variable-for/2187636#2187636
http://xion.io/post/code/python-all-wild-imports.html

We can still keep `from  captum.optim import loss ` but  we can define loss.L2 in the __all__. What do you think ?

Reply:
@NarineK I seem to be running into a weird __all__ bug. Here are 2 things I tried.

**Attempt 1**
```
from captum.optim import loss

__all__ = ["loss.L2"]
```
Running `from captum.optim import *` gives me a `ModuleNotFoundError: No module named 'captum.optim.loss'` error.

**Attempt 2**
```
from captum.optim import loss

__all__ = ["L2"]
```
Running `from captum.optim import *` gives me a `AttributeError: module 'captum.optim' has no attribute 'L2'` error.

---

So I tried reorganizing the structure of the module e.g. `loss` now has its own folder with its own `__init__.py` file that contains `__all__` declared for loss classes/functions. The public functions/classes are now declared in the `__all__` variables for the respective `__init__.py` files. Running `from captum.optim import *` works and prevents access to undeclared functions. Does this solution work?

Reply:
Thank you for trying out all those options, @greentfrapp! I think that the extra packages with one underscore file in it look a bit unusual. How about if we leave it as it was before?
```
from captum.optim import loss, transform
```
If there are any issues with public API or API documentation generation with sphinx we can look into them and make changes accordingly later.
Sorry for the confusion. I was thinking that it might be simple to define it. For now we can have `loss, transform` etc in the `__all__`.

Reply:
Hey Narine no worries! I just reverted it to the previous version.

Reply:
Thank you, @greentfapp! I came across this today.
https://github.com/fastai/fastai/blob/master/fastai/interpret.py#L3
This means that in the loss file for example we can define which classes do we want to expose and hopefully the API will pick that.

```
import *
__all__ in a module, e.g. module.py:

__all__ = ['foo', 'Bar']

means that when you import * from the module, only those names in the __all__ are imported:


```
https://stackoverflow.com/questions/44834/can-someone-explain-all-in-python

Reply:
Oh I tried that before but I realized that if I do `from captum.optim import *`, then I'll still be able to import whatever I want from `loss` or `optimization` etc. The `__all__` only applies if I do something like `from captum.optim._core.loss import *`.

But I added the `__all__` in `loss`, `optimization`, `images`, `transforms`, `reducer` and `circuits` in the latest commit! Since it might help with Sphinx and also act as an indication for users as to what is stable and not stable.

Reply:
Thank you, @greentfrapp! That sounds great!